target/linux/ramips/dts/DWR-116-A1.dts

```diff
@@ -0,0 +1,139 @@
+/dts-v1/;
+
+#include "mt7620n.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "dlink,dwr-116-a1", "ralink,mt7620n-soc";
+	model = "D-link DWR-116 A1";
+
+	gpio-keys-polled {
+	compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+
+		wps {
+			label = "wps";
+			gpios = <&gpio0 2 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+		reset {
+			label = "reset";
+			gpios = <&gpio0 1 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		lan1 {
+			label = "dwr-116-a1:green:lan1";
+			gpios = <&gpio2 0 GPIO_ACTIVE_LOW>;
+		};
+
+		lan2 {
+			label = "dwr-116-a1:green:lan2";
+			gpios = <&gpio2 1 GPIO_ACTIVE_LOW>;
+		};
+
+		lan3 {
+			label = "dwr-116-a1:green:lan3";
+			gpios = <&gpio2 2 GPIO_ACTIVE_LOW>;
+		};
+
+		lan4 {
+			label = "dwr-116-a1:green:lan4";
+			gpios = <&gpio2 3 GPIO_ACTIVE_LOW>;
+		};
+
+		/*wan {  /*WAN LED IS UART TX*/
+			/*label = "dwr-116-a1:green:wan";
+			gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
+		};*/
+
+		wifi {
+			label = "dwr-116-a1:green:wifi";
+			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+};
+
+&gpio1 {
+	status = "okay";
+};
+
+&gpio2 {
+	status = "okay";
+};
+
+&gpio3 {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <50000000>;
+
+		partition@0 {
+			label = "uboot";
+			reg = <0x0 0x40000>; /*todo sprawdzic*/
+			read-only;
+		};
+		
+		partition@40000 {
+			label = "uboot-env";
+			reg = <0x40000 0x10000>;
+			read-only;
+		};
+
+		partition@50000 {
+			label = "firmware";
+			reg = <0x50000 0x7a0000>;
+		};
+
+		config: partition@7f0000 {
+			label = "config";
+			reg = <0x7f0000 0x10000>;
+			read-only;
+		};
+	};
+};
+
+&ehci {
+	status = "okay";
+};
+
+&ohci {
+	status = "okay";
+};
+
+&ethernet {
+	mtd-mac-address = <&config 0xe291>;
+	mediatek,portmap = "llllw";
```

Comment:
if you add here:

```
	pinctrl-names = "default";
	pinctrl-0 = <&ephy_pins>;
```

you can let the hardware control the leds. It would allow you to drop the ucidef_set_led_switch from userspace. You need to drop the `ephy` from the gpio mux as well. Since it is a rather controversial discussion if LEDs should be controllable from userspace or handled by hardware, I only mention how it can be done. Handling LEDs by hardware would mean less cpu cycles wastes for userspace LED handling.

Reply:
@mkresin 
I agree with You.

All changes are done.